packages/ember-metal/lib/meta.js

```diff
@@ -100,77 +99,131 @@ function inheritedMap(name, Meta) {
   let key = memberProperty(name);
   let capitalized = capitalize(name);
 
-  Meta.prototype['writable' + capitalized] = function() {
-    return this._getOrCreateInheritedMap(key);
+  Meta.prototype['write' + capitalized] = function(subkey, value) {
+    let map = this._getOrCreateOwnMap(key);
+    map[subkey] = value;
   };
 
-  Meta.prototype['readable' + capitalized] = function() {
-    return this._getInherited(key);
+  Meta.prototype['peek' + capitalized] = function(subkey) {
+    return this._findInherited(key, subkey);
   };
 
-  Meta.prototype['peek' + capitalized] = function(subkey) {
-    let map = this._getInherited(key);
-    if (map) {
-      return map[subkey];
+  Meta.prototype['forEach' + capitalized] = function(fn) {
+    let pointer = this;
+    let seen = new EmptyObject();
+    while (pointer !== undefined) {
+      let map = pointer[key];
```

Comment:
then instead of keys + forEach, `for (var k in map) { }`, as long as we don't delete from the map. Will perform the best.